src/sage/algebras/quatalg/quaternion_algebra.py

```diff
@@ -3020,6 +3020,119 @@ def cyclic_right_subideals(self, p, alpha=None):
+            sage: I.is_integral()
+            True 
+            sage: I = R.ideal([1/2 + 2*j + 140*k, 2*i + 4*j + 150*k, 8*j + 104*k, 152*k])
+            sage: I.is_integral()
+            False
+
+        """
+        if self.__left_order is not None:
+            return all([b in self.left_order() for b in self.basis()])
+        elif self.__right_order is not None:
+            return all([b in self.right_order() for b in self.basis()])
+        else:
+            self_square = self**2
+            return all([b in self for b in self_square.basis()])
+
+
+    def primitive_decomposition(self):
+        r"""
+            Let `I` = ``self``. If `I` is an integral left `\mathcal{O}`-ideal return its decomposition as an equivalent primitive ideal and an integer such that their product is the initial ideal. 
+            Returns:
+
+                `J`: equivalent primitive ideal to `I`, i.e. equivalent ideal not contained in `n\mathcal{O}` for any `n>0`
+                `g`: smallest integer such that `I \subset g\mathcal{O}`
+
+            EXAMPLES: 
+                sage: A.<i,j,k> = QuaternionAlgebra(QQ, -1,-11)
+                sage: I = A.ideal([1/2 + 1/2*i + 1/2*j + 3/2*k, i + k, j + k, 2*k])
+                sage: I.primitive_decomposition()
+                (Fractional ideal (1/2 + 1/2*i + 1/2*j + 3/2*k, i + k, j + k, 2*k), 1)
+                sage: J = A.ideal([7/2 + 7/2*i + 49/2*j + 91/2*k, 7*i + 21*k, 35*j + 35*k, 70*k])
+                sage: Jequiv, g = J.primitive_decomposition()
+                sage: Jequiv*g == J 
+                True 
+                sage: Jequiv, g
+                (Fractional ideal (1/2 + 1/2*i + 7/2*j + 13/2*k, i + 3*k, 5*j + 5*k, 10*k), 7)
+
+            TESTS: 
+
+            Checks on random crafted ideals that they decompose as expected. 
+
+                sage: for d in ( m for m in range(400, 750) if is_squarefree(m) ):
+                ....:     A = QuaternionAlgebra(d)
+                ....:     O = A.maximal_order()
+                ....:     for _ in range(10):
+                ....:         a = O.random_element()
+                ....:         if not a.is_constant(): # avoids a = 0
+                ....:             I = a*O + a.reduced_norm()*O
+                ....:             if I.is_integral():
+                ....:                 J,g = I.primitive_decomposition()
+                ....:                 assert J*g == I
+                ....:                 assert J.is_primitive()
+
+        """
+
+        if not self.is_integral():
+            raise ValueError("primitive ideals are defined only for integral ideals")
+
+        I_basis = self.basis_matrix()
+        O_basis = self.left_order().basis_matrix()
+
+        # Write I in the basis of its left order via rref
+        M = block_matrix(1,2,[O_basis.transpose(),I_basis.transpose()]).rref()[:,4:]
+        g = Integer(gcd((gcd(M_row) for M_row in M)))
+
+        # If g is 1 then the ideal is primitive
+        if g == 1:
```

Comment:
The test `g.is_one()` is significantly faster than `g == 1`, since the latter goes through the type coercion system to find a common parent type to do the comparison in.